Sources/AsyncHTTPClient/RequestValidation.swift

```diff
@@ -16,93 +16,79 @@ import NIOCore
                 // anticipate such a body.
-                return metadata
+                break
             default:
                 // A user agent SHOULD send a Content-Length in a request message when
                 // no Transfer-Encoding is sent and the request method defines a meaning
                 // for an enclosed payload body.
                 self.add(name: "Content-Length", value: "0")
-                return metadata
             }
-        }
-
-        if case .TRACE = method {
-            // A client MUST NOT send a message body in a TRACE request.
-            // https://tools.ietf.org/html/rfc7230#section-4.3.8
-            throw HTTPClientError.traceRequestWithBody
-        }
-
-        guard (encodings.lazy.filter { $0 == "chunked" }.count <= 1) else {
-            throw HTTPClientError.chunkedSpecifiedMultipleTimes
-        }
-
-        if encodings.isEmpty {
-            if let length = body.length {
-                self.remove(name: "Content-Length")
-                contentLength = length
-            } else if !self.contains(name: "Content-Length") {
-                transferEncoding = "chunked"
+            return .init(connectionClose: connectionClose, body: .fixedSize(0))
+        case .fixed(let length):
+            if case .TRACE = method {
+                // A client MUST NOT send a message body in a TRACE request.
+                // https://tools.ietf.org/html/rfc7230#section-4.3.8
+                throw HTTPClientError.traceRequestWithBody
+            }
+            if encodings.isEmpty {
+                self.add(name: "Content-Length", value: String(length))
+                return .init(connectionClose: connectionClose, body: .fixedSize(length))
+            } else {
+                self.add(name: "Transfer-Encoding", value: encodings.joined(separator: ", "))
+                return .init(connectionClose: connectionClose, body: .stream)
+            }
+        case .dynamic:
+            if case .TRACE = method {
+                // A client MUST NOT send a message body in a TRACE request.
+                // https://tools.ietf.org/html/rfc7230#section-4.3.8
+                throw HTTPClientError.traceRequestWithBody
             }
-        } else {
-            self.remove(name: "Content-Length")
 
-            transferEncoding = encodings.joined(separator: ", ")
-            if !encodings.contains("chunked") {
-                guard let length = body.length else {
-                    throw HTTPClientError.contentLengthMissing
-                }
-                contentLength = length
+            if encodings.isEmpty && contentLength == nil {
+                // if a user forgot to specify a Content-Length and Transfer-Encoding, we will set it for them
+                self.add(name: "Transfer-Encoding", value: "chunked")
+            } else {
+                self.add(name: "Transfer-Encoding", value: encodings.joined(separator: ", "))
```

Comment:
We also run in this case if the user has set a Content-Length header explicitly. I think in the `dynamic` case we should trust a user set content length header (though we should validate it is a positive int).

A case might be streaming a file somewhere.

Reply:
I'm not really sure about this. A user can always specify the length of a stream upfront through our API and doesn't need to set the `Content-Length` header manually through the `length` parameter.
https://github.com/swift-server/async-http-client/blob/62ad8183765c23c48de67aa8ff08b653d679e0b4/Sources/AsyncHTTPClient/HTTPHandler.swift#L61-L69
This would give them a way to do the same thing in to different ways and they may conflict.
I would almost say that we should always ignore/remove "Content-Length" header.
The same is almost true for `Transport-Encoding` but I'm not fully understanding all use cases for this one.

Reply:
If I read this code snipped correctly, `Transport-Encoding` can only be `chunked` because SwiftNIO always replace the `Transport-Encoding` with `chunked` or doesn't do chunked encoding if it is not the only encoding. 
https://github.com/apple/swift-nio/blob/addf69cfe60376c325397c8926589415576b1dd1/Sources/NIOHTTP1/HTTPEncoder.swift#L104-L129

Reply:
> As to supporting Content-Length: yes, we should emit it, but we should do so when users tell us the size of their stream. We shouldn't tolerate them setting that header.

Okay, I'm happy with that!

In that case we need to fix the code comment though here:
https://github.com/swift-server/async-http-client/blob/62ad8183765c23c48de67aa8ff08b653d679e0b4/Sources/AsyncHTTPClient/HTTPHandler.swift#L61-L69

If no Content-Length header is set, we will just reach for chunked.